lib/ProductOpener/Products.pm

```diff
@@ -1022,54 +1023,93 @@ sub retrieve_product_rev ($product_id, $rev, $include_deleted = 0) {
+Utility function to change the barcode of a product.
+Fails and returns an error if the code is invalid, or if there is already a product with the new code.
+
+=head3 Parameters
+
+=head4 $product_ref
+
+=head4 $new_code
+
+=head3 Return value
+
+If successful: undef
+If there was an error: invalid_code or new_code_already_exists
+
+=cut
+
+sub change_product_code ($product_ref, $new_code) {
+
+	# Currently only called by admins and moderators
 
 	my $code = $product_ref->{code};
-	my $new_server = "";
-	my $new_data_root = $data_root;
-
-	if ($new_code =~ /^([a-z]+)$/) {
-		$new_server = $1;
-		if (    (defined $options{other_servers})
-			and (defined $options{other_servers}{$new_server})
-			and ($options{other_servers}{$new_server}{data_root} ne $data_root))
-		{
-			$new_code = $code;
-			$new_data_root = $options{other_servers}{$new_server}{data_root};
-		}
-	}
 
 	$new_code = normalize_code($new_code);
 	if (not is_valid_code($new_code)) {
-		push @$errors_ref, lang("invalid_barcode");
+		return "invalid_code";
 	}
 	else {
 		# check that the new code is available
-		if (-e "$new_data_root/products/" . product_path_from_id($new_code)) {
-			push @{$errors_ref}, lang("error_new_code_already_exists");
-			$log->warn(
-				"cannot change product code, because the new code already exists",
-				{code => $code, new_code => $new_code, new_server => $new_server}
-			) if $log->is_warn();
+		if (-e "$data_root/products/" . product_path_from_id($new_code) . "/product.sto") {
+			$log->warn("cannot change product code, because the new code already exists",
+				{code => $code, new_code => $new_code})
+				if $log->is_warn();
+			return "error_new_code_already_exists";
 		}
 		else {
 			$product_ref->{old_code} = $code;
 			$code = $new_code;
 			$product_ref->{code} = $code;
-			if ($new_server ne '') {
-				$product_ref->{new_server} = $new_server;
-			}
-			$log->info("changing code",
-				{old_code => $product_ref->{old_code}, code => $code, new_server => $new_server})
+			$log->info("changing code", {old_code => $product_ref->{old_code}, code => $code})
 				if $log->is_info();
 		}
 	}
 
 	return;
 }
 
+=head2 change_product_type ($product_ref, $new_product_type)
+
+Utility function to change the product type of a product.
+Fails and returns an error if the product type is invalid.
+
+=head3 Parameters
+
+=head4 $product_ref
+
+=head4 $new_product_type
+
+=head3 Return value
+
+If successful: undef
+If there was an error: invalid_product_type
+
+=cut
+
+sub change_product_type ($product_ref, $new_product_type) {
+
+	# Currently only called by admins and moderators
+
+	my $product_type = $product_ref->{product_type};
+
+	# Return if the product type is already the new product type, or if the new product type is not defined
+	if ((not defined $new_product_type) or ((not defined $options{product_types_flavors}{$new_product_type}))) {
+		return "invalid_product_type";
+	}
+	elsif ($product_type ne $new_product_type) {
```

Comment:
shan't we at least emit a warning if the new product type is the same as the old one ?

Reply:
From the API point of view, it just sends product_type like it could send product_name_en, we don't return warnings if the product name is already the same. In practice change_product_type() is called only if the product_type is different.